### PR TITLE
Preserve image attributes when cropping a custom logo image

### PR DESCRIPTION
**Pull request by anton-vlasenko**
This PR aims to implement preserving attributes when cropping a custom logo image.

Steps to test this PR: https://core.trac.wordpress.org/ticket/37750#comment:26

Trac ticket: https://core.trac.wordpress.org/ticket/37750

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: src/wp-admin/includes/ajax-actions.php
Comment (dream-encode) [inline src/wp-admin/includes/ajax-actions.php]:
```suggestion
			$has_title           = mb_strlen( trim( $original_attachment->post_title ) ) > 0;
```
A Yoda condition here degrades readability, per the WPCS(https://developer.wordpress.org/coding-standards/wordpress-coding-standards/php/)

File: src/wp-admin/includes/ajax-actions.php
Comment (dream-encode) [inline src/wp-admin/includes/ajax-actions.php]:
```suggestion
			$has_description     = mb_strlen( trim( $original_attachment->post_content ) ) > 0;
```
A Yoda condition here degrades readability, per the WPCS(https://developer.wordpress.org/coding-standards/wordpress-coding-standards/php/)

File: tests/phpunit/tests/ajax/CropImage.php
Comment (dream-encode) [inline tests/phpunit/tests/ajax/CropImage.php]: There's a few assertions here in the same test.  In this instance, it's best to include the `$message` parameter when you have multiple assertions in a test.  See https://make.wordpress.org/core/handbook/testing/automated-testing/writing-phpunit-tests/#using-assertions

File: tests/phpunit/tests/ajax/CropImage.php
Comment (dream-encode) [inline tests/phpunit/tests/ajax/CropImage.php]: Should probably note `@ticket 37750` here as well.

Comment (dream-encode) [inline tests/phpunit/tests/ajax/CropImage.php]:
```suggestion
	 * @covers ::wp_ajax_crop_image()
```

File: tests/phpunit/tests/ajax/CropImage.php
Comment (dream-encode) [inline tests/phpunit/tests/ajax/CropImage.php]:
```suggestion
	 * @covers ::wp_ajax_crop_image()
```

**Review by dream-encode — COMMENTED** [review]
More to address here in the actual logic, but these are the things that jump out at me straight away.

File: src/wp-admin/includes/ajax-actions.php
Comment (anton-vlasenko) [reply src/wp-admin/includes/ajax-actions.php]: Thank you for the feedback! Fixed in fdaa748a589a8cb0ab7e21b624e38f35284027df

File: src/wp-admin/includes/ajax-actions.php
Comment (anton-vlasenko) [reply src/wp-admin/includes/ajax-actions.php]: Fixed in https://github.com/WordPress/wordpress-develop/commit/fdaa748a589a8cb0ab7e21b624e38f35284027df

File: tests/phpunit/tests/ajax/CropImage.php
Comment (anton-vlasenko) [reply tests/phpunit/tests/ajax/CropImage.php]: Thank you! Fixed in 4a548660707894c1983fae48939ce1af07d98d78

Comment (anton-vlasenko) [reply tests/phpunit/tests/ajax/CropImage.php]: Thank you for the feedback.
Fixed in 4a548660707894c1983fae48939ce1af07d98d78.
I wouldn't use parenthesis at the end because it seems they are not needed according to the PHPUnit documentation.

File: tests/phpunit/tests/ajax/CropImage.php
Comment (anton-vlasenko) [reply tests/phpunit/tests/ajax/CropImage.php]: Fixed in 4a548660707894c1983fae48939ce1af07d98d78

File: tests/phpunit/tests/ajax/CropImage.php
Comment (anton-vlasenko) [reply tests/phpunit/tests/ajax/CropImage.php]: I agree. Thank you for the feedback.
Fixed in f96f9d31420ce77768c94bddfeb9babedafaa716

File: tests/phpunit/tests/ajax/CropImage.php
Comment (hellofromtonya) [reply tests/phpunit/tests/ajax/CropImage.php]: Yes, I agree with not including the `()` at the end of the `@covers` annotation.  https://github.com/WordPress/wordpress-develop/commit/4a548660707894c1983fae48939ce1af07d98d78 LGTM 👍

File: src/wp-admin/includes/ajax-actions.php
Comment (hellofromtonya) [inline src/wp-admin/includes/ajax-actions.php]: A simplification (see it in action https://3v4l.org/hCPCX):

```php
( '' !== trim( $original_attachment->post_title ) );
```

```suggestion
			$has_title           = ( '' !== trim( $original_attachment->post_title ) );
			$has_description     = ( '' !== trim( $original_attachment->post_content ) );
```

File: src/wp-admin/includes/ajax-actions.php
Comment (hellofromtonya) [inline src/wp-admin/includes/ajax-actions.php]: Same here
```suggestion
			if ( '' !== trim( $original_attachment->post_excerpt ) ) {
```

File: src/wp-admin/includes/ajax-actions.php
Comment (hellofromtonya) [inline src/wp-admin/includes/ajax-actions.php]:
```suggestion
			if ( '' !== trim( $original_attachment->_wp_attachment_image_alt ) ) {
```

File: tests/phpunit/tests/ajax/CropImage.php
Comment (hellofromtonya) [inline tests/phpunit/tests/ajax/CropImage.php]: As this test class is specific for one function, the `@covers` annotation can be set at the class level rather than at the test method level.

```suggestion
 * @group ajax
 * @covers ::wp_ajax_crop_image
```

File: tests/phpunit/tests/ajax/CropImage.php
Comment (hellofromtonya) [inline tests/phpunit/tests/ajax/CropImage.php]: As both tests use this, I'd suggest moving it to `set_up()` fixture method instead as then the tests are setup and ready to go.

File: tests/phpunit/tests/ajax/CropImage.php
Comment (hellofromtonya) [inline tests/phpunit/tests/ajax/CropImage.php]:
```suggestion
```

Core doesn't use `@return void`.

File: tests/phpunit/tests/ajax/CropImage.php
Comment (hellofromtonya) [inline tests/phpunit/tests/ajax/CropImage.php]:
```suggestion
```

Core doesn't use `@return void`.

File: tests/phpunit/tests/ajax/CropImage.php
Comment (hellofromtonya) [inline tests/phpunit/tests/ajax/CropImage.php]: To avoid duplication and inconsistencies, I'd suggest instead to use `$this->_make_attachment()` helper method (which is available through inheritance coming from `WP_UnitTestCase_Base`). If `$with_metadata` is `true`, then update the post and its metadata. Doing this ensures the logic needed to create an attachment is centralized in one location.

File: tests/phpunit/tests/ajax/CropImage.php
Comment (hellofromtonya) [inline tests/phpunit/tests/ajax/CropImage.php]: The class name should follow this convention `Tests_[group]_FunctionName`, which changes this test class to `Tests_Ajax_WpAjaxCropImage`.

and the file should be camelCase of the function name, e.g. `wpAjaxCropImage.php`.

File: tests/phpunit/tests/ajax/CropImage.php
Comment (hellofromtonya) [inline tests/phpunit/tests/ajax/CropImage.php]: This logic will not be reached if a test fails. Instead move it to `tear_down()` which means the ID of each will need to become a property for `tear_down()` to have access.

File: tests/phpunit/tests/ajax/CropImage.php
Comment (hellofromtonya) [inline tests/phpunit/tests/ajax/CropImage.php]: same note as previous test.

**Review by hellofromtonya — CHANGES_REQUESTED** [review]
Overall looks good. Some changes for consistency.

File: tests/phpunit/tests/ajax/CropImage.php
Comment (anton-vlasenko) [reply tests/phpunit/tests/ajax/CropImage.php]: I've already tried to move this code to `tear_down`, and the attached images weren't deleted. I've just realised (thanks to your comment) that I can fix it by calling `parent::tear_down()` at the very end of my `tear_down` method. 🤦

Otherwise, DB entities will be rolled back in the parent `tear_down` method, and the `wp_delete_attachment` method will not work.
Fixed in c55674c7b8e71c6043f2bd6aa35cf42d5c6d51be

File: tests/phpunit/tests/ajax/CropImage.php
Comment (anton-vlasenko) [reply tests/phpunit/tests/ajax/CropImage.php]: Fixed in https://github.com/WordPress/wordpress-develop/commit/c55674c7b8e71c6043f2bd6aa35cf42d5c6d51be

File: tests/phpunit/tests/ajax/CropImage.php
Comment (anton-vlasenko) [reply tests/phpunit/tests/ajax/CropImage.php]: Fixed in 412e86234bf99b738912fecd5d84be4897a4eb30

File: tests/phpunit/tests/ajax/CropImage.php
Comment (anton-vlasenko) [reply tests/phpunit/tests/ajax/CropImage.php]: Good to know. Thank you.
Fixed in 412e86234bf99b738912fecd5d84be4897a4eb30

File: tests/phpunit/tests/ajax/CropImage.php
Comment (anton-vlasenko) [reply tests/phpunit/tests/ajax/CropImage.php]: That's good to know! Thank you.
Fixed in 743688eec4566886bdfac2f9c0312e224eb19bc0
P.S. It seems that all classes/files in the `tests/phpunit/tests/ajax` folder have to be renamed because they don't follow this convention.

File: src/wp-admin/includes/ajax-actions.php
Comment (anton-vlasenko) [reply src/wp-admin/includes/ajax-actions.php]: I absolutely agree. It's more efficient to compare values with an empty string.
Fixed in a2e8c8690e82677dfeaea998b027b10f72e6cbdd

File: src/wp-admin/includes/ajax-actions.php
Comment (anton-vlasenko) [reply src/wp-admin/includes/ajax-actions.php]: Fixed in https://github.com/WordPress/wordpress-develop/commit/a2e8c8690e82677dfeaea998b027b10f72e6cbdd

File: src/wp-admin/includes/ajax-actions.php
Comment (anton-vlasenko) [reply src/wp-admin/includes/ajax-actions.php]: Fixed in https://github.com/WordPress/wordpress-develop/commit/a2e8c8690e82677dfeaea998b027b10f72e6cbdd

File: tests/phpunit/tests/ajax/CropImage.php
Comment (anton-vlasenko) [reply tests/phpunit/tests/ajax/CropImage.php]: > I'd suggest instead to use $this->_make_attachment() helper method 

That's good to know. Fixed in a39a55141d5da872225890a2054722f567a657de

File: tests/phpunit/tests/ajax/CropImage.php
Comment (anton-vlasenko) [reply tests/phpunit/tests/ajax/CropImage.php]: Fixed in a4cb089d043fefcc7e957e13428f4e0da059ffbd

File: tests/phpunit/tests/ajax/CropImage.php
Comment (anton-vlasenko) [reply tests/phpunit/tests/ajax/CropImage.php]: Fixed in a4cb089d043fefcc7e957e13428f4e0da059ffbd

File: tests/phpunit/tests/ajax/CropImage.php
Comment (hellofromtonya) [reply tests/phpunit/tests/ajax/CropImage.php]: >P.S. It seems that all classes/files in the `tests/phpunit/tests/ajax` folder have to be renamed because they don't follow this convention.

This is true throughout the test suite. Can be remedied as part of https://core.trac.wordpress.org/ticket/54725. Feel free to create a PR.

File: src/wp-admin/includes/ajax-actions.php
Comment (hellofromtonya) [inline src/wp-admin/includes/ajax-actions.php]: Instead of assigning the `post_content` above and then overwriting it, move this inline into the array above.

Comment (hellofromtonya) [reply src/wp-admin/includes/ajax-actions.php]: Currently updating and will push shortly

File: src/wp-admin/includes/ajax-actions.php
Comment (hellofromtonya) [inline src/wp-admin/includes/ajax-actions.php]: @anton-vlasenko I added this check to only use the original image's title _if it was edited_. When there's no title in the image's metadata, by default the image's filename is set as the `post_title`. This check is making sure the `post_title` is not the same as that default behavior. If no, then it will use the original image's title.

Comment (hellofromtonya) [reply src/wp-admin/includes/ajax-actions.php]: The tests will need updating though.

File: src/wp-admin/includes/ajax-actions.php
Comment (hellofromtonya) [inline src/wp-admin/includes/ajax-actions.php:3992]: @anton-vlasenko I moved this back here for consistency and then renamed to make it more clear _what_ the value is.

File: src/wp-admin/includes/ajax-actions.php
Comment (anton-vlasenko) [reply src/wp-admin/includes/ajax-actions.php]: Thank you!
I'm working on updating the tests.

File: src/wp-admin/includes/ajax-actions.php
Comment (anton-vlasenko) [reply src/wp-admin/includes/ajax-actions.php:3992]: Thank you, @hellofromtonya.

File: src/wp-admin/includes/ajax-actions.php
Comment (hellofromtonya) [inline src/wp-admin/includes/ajax-actions.php]:
```suggestion
				/*
				 * Check if the original image has a title other than the "filename" default,
				 * meaning the image had a title when originally uploaded or its title was edited.
				 */
```

File: src/wp-admin/includes/ajax-actions.php
Comment (hellofromtonya) [inline src/wp-admin/includes/ajax-actions.php]:
```suggestion
				( '' !== trim( $original_attachment->post_title ) ) &&
```
for consistency with the other checks.

File: tests/phpunit/tests/ajax/wpAjaxCropImage.php
Comment (hellofromtonya) [inline tests/phpunit/tests/ajax/wpAjaxCropImage.php]:
```suggestion
```

**Review by hellofromtonya — APPROVED** [review]
The implementation and tests LGTM 👍
Approving - pending input from @felixarntz ([see context in the Trac ticket comment](https://core.trac.wordpress.org/ticket/37750#comment:29)).

File: src/wp-admin/includes/ajax-actions.php
Comment (felixarntz) [inline src/wp-admin/includes/ajax-actions.php:4005]: Bit of a nit-pick, but I think it would make sense to include this immediately in the associative array above, right after `post_title` and `post_content`, which would clarify the relationship of the three when reading the code. The default for this would need to be the empty string (`''`), which is the default for `wp_insert_post` (and thus `wp_insert_attachment`).

File: src/wp-admin/includes/ajax-actions.php
Comment (felixarntz) [inline src/wp-admin/includes/ajax-actions.php]: Similarly to my above comment, we could integrate this into the associative array passed to `wp_insert_attachment` above, which would colocate it with the other other related data. Other than in my previous comment though, here it would be more about avoiding an extra call to `update_post_meta`.

We can use the `meta_input` key (supported by `wp_insert_post` and thus also by `wp_insert_attachment`) to pass this information, basically like `$object['meta_input'] = array( '_wp_attachment_image_alt' => wp_slash( $original_attachment->_wp_attachment_image_alt ) )`.

**Review by felixarntz — APPROVED** [review]
This looks great @anton-vlasenko! I left two very minor comments for your consideration, although they certainly don't need to block this from getting merged. Please take a look.

@hellofromtonya I've left a comment in https://core.trac.wordpress.org/ticket/37750#comment:33 as well.

File: src/wp-admin/includes/ajax-actions.php
Comment (anton-vlasenko) [reply src/wp-admin/includes/ajax-actions.php:4005]: Thank you for the feedback.
I intentionally didn't include it in the associative array.
Why? I prefer not to specify a default value if the task does not require that.
I prefer to let `wp_insert_attachment` decide what value to use as default.

File: src/wp-admin/includes/ajax-actions.php
Comment (anton-vlasenko) [reply src/wp-admin/includes/ajax-actions.php]: I absolutely agree. I didn't know `wp_insert_post` supports the `meta_input` key.
Fixed in 9e2a262b7e01d26600d62375f2bd16536e7e931e
cc @hellofromtonya 
I've implemented the change suggested in @felixarntz`s [comment](https://github.com/WordPress/wordpress-develop/pull/2441#discussion_r834770715) above. I tested it manually and it worked as intended. Unit tests also passed.
But if that change will require retesting all use cases manually, please let me know, and I will revert that change.
I also don't want to block this PR from getting merged.

File: src/wp-admin/includes/ajax-actions.php
Comment (felixarntz) [reply src/wp-admin/includes/ajax-actions.php:4005]: That's a fair point. Works for me 👍

**Comment by hellofromtonya** [comment]
Committed via changeset https://core.trac.wordpress.org/changeset/53027.

**Pull request closed without merging**
